lib/oauth/authenticator.js

```diff
@@ -21,31 +31,37 @@ OAuthAuthenticator.prototype.withLocalValidation = function withLocalValidation(
   return this;
 };
 
-OAuthAuthenticator.prototype.authenticate = function authenticate(req,callback){
+OAuthAuthenticator.prototype.authenticate = function authenticate(req, callback){
   var self = this;
-  var done;
-  if(typeof req==='object' && typeof callback === 'function'){
-    done = callback;
-  }else{
+
+  if (typeof req !== 'object' || typeof callback !== 'function') {
     throw new Error('authenticate must be called with a request object and callback function');
   }
-  var authHeader = req && req.headers && req.headers.authorization;
 
-  if (authHeader){
-    if(authHeader.match(/Bearer/i)){
-      var parts = authHeader.split(' ');
-      var token = parts[parts.length-1];
-      new JwtAuthenticator(self.application).authenticate(token,done);
-    }else{
-      done(self.unauthenticated());
+  var authenticator = null;
+  var token = retrieveAuthTokenFromRequest(req);
+
+  if (token) {
+    authenticator = new JwtAuthenticator(self.application);
+
+    if (this.localValidation) {
+      authenticator.withLocalValidation();
     }
-  }else if (req.body && req.body.grant_type==='password'){
-    new OAuthPasswordGrantRequestAuthenticator(self.application).authenticate(req.body,callback);
-  }else if (req.body && req.body.grant_type==='refresh_token'){
-    new OAuthRefreshTokenGrantRequestAuthenticator(self.application).authenticate(req.body,callback);
-  }else{
-    done(self.unauthenticated());
+  } else if (req.body) {
+    token = req.body;
+    if (req.body.grant_type === 'password') {
```

Comment:
I think it's easier to follow the logic of these lines rather than with the long `if-else` lines:

```javascript
switch (req.body.grant_type) {
  case 'password':
    authenticator = new OAuthPasswordGrantRequestAuthenticator(self.application);
    break;

  case 'refresh_token':
    authenticator = new OAuthRefreshTokenGrantRequestAuthenticator(self.application);
    break;
}
```

**vs**

```javascript
if (req.body.grant_type === 'password') {
  authenticator = new OAuthPasswordGrantRequestAuthenticator(self.application);
} else if (req.body.grant_type === 'refresh_token') {
  authenticator = new OAuthRefreshTokenGrantRequestAuthenticator(self.application);
}
```